src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineControl.xaml.cs

```diff
@@ -0,0 +1,388 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Windows;
+using System.Windows.Controls;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.VisualStudio.Editor;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.VisualStudio.LanguageServices.Implementation;
+using Microsoft.VisualStudio.LanguageServices.Implementation.LanguageServiceBrokerShim;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.TextManager.Interop;
+using Microsoft.VisualStudio.Threading;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices
+{
+    /// <summary>
+    /// Interaction logic for DocumentOutlineControl.xaml
+    /// </summary>
+    internal partial class DocumentOutlineControl : UserControl, IVsCodeWindowEvents
+    {
+        public IVsEditorAdaptersFactoryService EditorAdaptersFactoryService { get; }
+        public IVsCodeWindow CodeWindow { get; }
+
+        private IThreadingContext ThreadingContext { get; }
+        private ILanguageServiceBrokerShim LanguageServiceBroker { get; }
+
+        /// <summary>
+        /// The type of sorting applied to the document model.
+        /// </summary>
+        private SortOption SortOption { get; set; }
+
+        /// <summary>
+        /// Queue to batch up work to do to get the current document model. Used so we can batch up a lot of events 
+        /// and only fetch the model once for every batch.
+        /// </summary>
+        private readonly AsyncBatchingWorkQueue _getModelQueue;
+
+        /// <summary>
+        /// Queue to batch up work to do to update the current document model and UI.
+        /// </summary>
+        private readonly AsyncBatchingWorkQueue _updateModelQueue;
+
+        /// <summary>
+        /// Queue to batch up work to do to highlight the currently selected symbol node.
+        /// </summary>
+        private readonly AsyncBatchingWorkQueue _highlightNodeQueue;
+
+        /// <summary>
+        /// The text snapshot from when the document symbol request was made.
+        /// </summary>
+        private ITextSnapshot? LspSnapshot { get; set; }
+
+        /// <summary>
+        /// Keeps track of the current primary and secondary text views.
+        /// </summary>
+        private readonly Dictionary<IVsTextView, ITextView> _trackedTextViews = new();
```

Comment:
doc what threads are allowed to access this.  if multiple threads can access doc how they do so safely.